PKI External CA DNS feature updates (#2656)
hashicorp/web-unified-docs: content/vault/v2.x/content/docs/secrets/pki-external-ca/index.mdx

$ vault write pki-external-ca/role/web-servers @role.json
```

## DNS providers for automatic challenge fulfillment

The PKI External CA secrets engine integrates with certain DNS providers to enable automatic
DNS-01 challenge fulfillment. You configure DNS providers globally and share them across all ACME accounts,
allowing you to automate certificate issuance without manual intervention or external systems.

### Challenge fulfillment modes

The plugin supports three challenge fulfillment modes:

1. **Manual/External Fulfillment** (default behavior): The plugin exposes challenges via API endpoints for

> external systems to fulfill. They retrieve challenge tokens from Vault and manually create DNS records or HTTP tokens.

Reviewer: ```suggestion
external systems to fulfill. The external systems retrieve challenge tokens from Vault and manually create DNS records or HTTP tokens.
```

1. **Automatic DNS Fulfillment**: Vault automatically manages DNS records using configured provider
credentials. When DNS providers are configured and match all the requested identifiers within an order, Vault handles

> the entire ACME certificate request process.

Reviewer: ```suggestion
1. **Automatic DNS Fulfillment**: Vault automatically manages DNS records
   using configured provider credentials. Vault handles the entire ACME
   certificate request process when it receives an order that matches all
   the requested identifiers for a configured DNS provider.
```
Style correction: write in active voice, make the sentence subject obvious

1. **Hybrid Fulfillment**: Some challenges are fulfilled manually while others are automatically fulfilled by Vault.

> This is useful when you have DNS providers configured for some domains but not others.

Reviewer: ```suggestion
1. **Hybrid Fulfillment**: You can configure Vault to fulfill some of
   the challenges automatically while others require manual fulfillment.
   Hybrid fulfillment is useful when you have DNS providers configured
   for some domains but not others.
```
Style correction: write in active voice

### DNS provider configuration

You manage DNS providers independently `config/dns/<provider-type>/<name>` and specify supported identifiers

> using glob patterns. During ACME challenge evaluation for any order from any ACME account, the system will:

Reviewer: ```suggestion
You manage DNS providers independently with the
`config/dns/<provider-type>/<name>` endpoint and specify the supported
identifiers using glob patterns. During ACME challenge evaluation for
any order from any ACME account, Vault:
```

> 1. Select one or more matching DNS providers based on the order's identifiers

Reviewer: ```suggestion
1. Selects one or more matching DNS providers based on the identifiers
   in the order.
```
Style correction: avoid possessives

> 1. Challenges with no matching provider require manual fulfillment

Reviewer: ```suggestion
1. Requires manual fulfillment for challenges with no matching provider.
```
We've introduced the list as actions the system takes. Each line item needs to be written so it makes sense for it to start with "Vault..."

> 1. Matching challenges will be automatically fulfilled using the DNS provider

Reviewer: ```suggestion
1. Identifies challenges for automatic fulfillment using the DNS provider.
```

> 1. In hybrid scenarios, manual challenges must be fulfilled first, then Vault performs automatic DNS updates

Reviewer: ```suggestion
1. Waits for manual challenge fulfillment then performs automatic DNS
   updates for challenges that qualify for automatic fulfillment.
```

### Supported DNS providers

The plugin supports multiple DNS providers, including:

> - **AWS Route53** - Amazon's DNS service

Reviewer: ```suggestion
- **AWS Route53** - Amazon DNS service
```

    access_key_id="AKIAIOSFODNN7EXAMPLE" \
    secret_access_key="wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY" \
    region="us-east-1"
```

</Tab>

<Tab heading="Azure DNS">

Configure Azure DNS as a DNS provider:

```shell-session
$ vault write pki-external-ca/config/dns/azure-dns/production \
    identifiers="*.example.com" \
    zone_name="my-zone-id" \
    client_id="00000000-0000-0000-0000-000000000000" \
    client_secret="your-client-secret"
```

</Tab>

<Tab heading="Google Cloud DNS">

Configure Google Cloud DNS as a DNS provider:

```shell-session
$ vault write pki-external-ca/config/dns/google-cloud-dns/production \
    identifiers="*.example.com" \
    credentials=@gcp-service-account.json
```

</Tab>

<Tab heading="RFC2136">

Configure RFC2136 for dynamic DNS updates:

```shell-session
$ vault write pki-external-ca/config/dns/rfc2136/bind-dns \
    identifiers="*.example.com" \
    nameserver="ns1.example.com:53" \
    tsig_key_name="my-key" \
    tsig_secret="base64-encoded-secret" \
    tsig_algorithm="hmac-sha256"
```

</Tab>

</Tabs>

### Testing DNS providers

Before using a DNS provider within an ACME workflow, a test API endpoint

> allows you to validate the DNS provider configuration can be used without requesting any certificates.

Reviewer: ```suggestion
You can use the test API endpoint before using a DNS provider within
an ACME workflow to confirm the DNS provider configuration works
without requesting any certificates.
```

> 

Reviewer: ```suggestion
For example, the following command confirms that:

- The DNS provider credentials are correct.
- Vault can create and delete DNS records.
- The DNS provider can service the specified identifier.

```

```shell-session
$ vault write pki-external-ca/dns/test/workflow \
    provider_type="aws-route53" \
    provider_name="production" \
    identifier="test.example.com"
```

This validates that:
- The DNS provider credentials are correct
- Vault can create and delete DNS records

> - The DNS provider can service the specified identifier

Reviewer: ```suggestion
```
Style correction: explain example code before presenting it, not after

   $ vault read pki-external-ca/role/web-servers/order/01936d8e-7c3a-7890-b123-456789abcdef/status
   ```

> 1. **If DNS providers are configured**: Vault automatically handles DNS-01 challenges for matching identifiers. Skip to step 5.

Reviewer: ```suggestion
1. **If you have DNS providers configured**, Vault automatically handles
   DNS-01 challenges for matching identifiers. Skip to step 5.
```

> 1. **If no DNS providers are configured or identifiers don't match**: Fetch the challenge token and auth key for each identifier from Vault:

Reviewer: ```suggestion
1. **If you do not have a DNS provider configured or the identifiers
   do not match**, fetch the challenge token and auth key for each
   identifier from Vault:
```

   $ vault read pki-external-ca/role/web-servers/order/01936d8e-7c3a-7890-b123-456789abcdef/status
   ```

> 1. **If DNS providers are configured**: Vault automatically handles DNS-01 challenges for matching identifiers. Skip to step 5.

Reviewer: ```suggestion
1. **If you have DNS providers configured**, Vault automatically handles
   DNS-01 challenges for matching identifiers. Skip to step 5.
```

> 1. **If no DNS providers are configured or identifiers don't match**: Fetch the challenge token and auth key for each identifier from Vault:

Reviewer: ```suggestion
1. **If you do not have a DNS provider configured or the identifiers
   do not match**, fetch the challenge token and auth key for each
   identifier from Vault:
```
